core/unix/signal_linux.c

```diff
@@ -364,6 +364,8 @@ handle_pre_extended_syscall_sigmasks(dcontext_t *dcontext, kernel_sigset_t *sigm
     if (sizemask != sizeof(kernel_sigset_t))
         return false;
     ASSERT(!info->pre_syscall_app_sigprocmask_valid);
+    if (sigmask == NULL)
+        return true;
```

Comment:
I think that's almost the same to me. The only code it is running unnecessarily is setting the app's original parameter to NULL (which it already is). I actually thought it's cleaner like this. What about adding a comment explaining this?

Reply:
It's checking sig_pending, which seems fragile because it is not clear whether handle_pre_extended_syscall_sigmasks (not sure why it's plural "masks"?) is setting sig_pending just based on some general check or conditionally based on the sigmask: i.e., there's behavior split between os.c and here, but only here is checking for the NULL.  When reviewing the code, there are more cases to check for doing the right thing with this approach than with bailing early, and the worry is that new code will be added in os.c that forgets to handle NULL.

The pselect6 case which modifies the app's parameter by writing to memory has a corner case where an app that passed that parameter in read-only memory will work natively, but not under DR.  Avoiding that app-writing code makes DR slightly more transparent by matching native behavior more closely.   A pathological case, yes, but just showing that taking fewer actions is often more robust.

In any case, if left like this I would suggest a comment in the 3 os.c handlers stating that all of the code has to handle a NULL sigmask and must be a nop because no action should be taken for a NULL sigmask.

Reply:
> The pselect6 case which modifies the app's parameter by writing to memory has a corner case where an app that passed that parameter in read-only memory will work natively

Far-fetched but true.

Reply:
Done.